### PR TITLE
Aliasing lightning

### PR DESCRIPTION
**Pull request by eiyu**
Hi, I think I've make a progress. Could you checkout my work. thanks

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -29,22 +31,25 @@ const sendUAEvent = (event, search) => {
...
+  // console.log("current data ->", currentData);
   root.weapons.forEach(({ id, name }) => {
     weaponHash[id] = name;
   });
-
+  console.log(aliasingLightning(allItems[0]))
   allItems.forEach(item => {
-    addItem(item);
+    // change all the lightning to thunder
+    addItem(aliasingLightning(item));
```
Comment (eiyu) [inline index.js]: when you're adding up item the aliasingLightning function will be invoked with item as argument,
the aliasing function will return a copy object of item that has 'lightning' word in its factors property and replace the 'lightning' word to 'thunder'

File: package.json [hunk 2]
```diff
@@ -4,7 +4,7 @@
   "description": "",
   "main": "index.js",
   "scripts": {
-    "start": "node index.js"
+    "start": "nodemon index.js"
```
Comment (eiyu) [inline package.json]: I forgot to change the script, sorry. I'm kinda rushing when making pull request

**Review by eiyu — COMMENTED** [review]
There is something i forgot in package.json, because I'm using nodemon for development, I forgot to change the scripts back to "scripts" : {"start": node index.js}

File: commands/item.js [hunk 3]
```diff
@@ -90,8 +89,9 @@ const itemd = (client, msg, args, opts) => {
 const items = (client, msg, args, { region }) => {
   const allResults = [];
 
+  // console.log(getItemSet().length(), getSearchSet().length());
```
Comment (seiyria) [inline commands/item.js]: just remove this - console.log is not needed

File: commands/item.js [hunk 4]
```diff
@@ -90,8 +89,9 @@ const itemd = (client, msg, args, opts) => {
...
   const terms = args.split(',').map(x => x.trim().toLowerCase());
-
+  // console.log("terms is --> ",terms);
```
Comment (seiyria) [inline commands/item.js]: remove this

File: index.js [hunk 5]
```diff
@@ -1,10 +1,11 @@
 
+
 const axios = require('axios');
 const ua = require('universal-analytics');
 const Discord = require('discord.js');
 const Snoowrap = require('snoowrap');
 const { CommentStream } = require('snoostorm');
-
+const {aliasingLightning} = require('./utils/aliasingLightning')
```
Comment (seiyria) [inline index.js]: fix spacing, and name, and function location -

- should be in shared.js
- call it `aliasLightning` (this tense makes more sense)
- follow spacing of other imports

File: package.json [hunk 6]
```diff
@@ -4,7 +4,8 @@
   "description": "",
   "main": "index.js",
   "scripts": {
-    "start": "node index.js"
+    "start": "node index.js",
+    "dev": "nodemon index.js"
```
Comment (seiyria) [inline package.json]: Change this back

File: package.json [hunk 7]
```diff
@@ -31,5 +32,8 @@
     "lodash.uniqby": "^4.7.0",
     "snoostorm": "^1.0.2",
     "universal-analytics": "^0.4.20"
+  },
+  "devDependencies": {
+    "nodemon": "^1.19.3"
```
Comment (seiyria) [inline package.json]: Remove this

File: utils/aliasingLightning.js [hunk 8]
```diff
@@ -0,0 +1,100 @@
+// import I * from './item'
```
Comment (seiyria) [inline utils/aliasingLightning.js]: Get rid of all of these comments - we have this data format already so it does not need to be here: https://github.com/star-ocean-anamnesiac/anamnesiac/blob/master/src/assets/data/item/weapon/arm.yml

File: utils/aliasingLightning.js [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+//      { desc: '80% chance of surviving a lethal attack when at 10% of max HP or more',
+//        lb: 5 } ],
+//   obtained: 'Gacha',
+//   notes: 'Lightning imbue arms and as I type this the only elemental arms you can get. A must keep, even if you don\'t have HMillie these will come in useful later.\n',
+//   type: 'weapon',
+//   subtype: 'arm'
+// }
+// ]
+
+const containsLightning = element => {
```
Comment (seiyria) [inline utils/aliasingLightning.js]: use semicolons for your code

File: utils/aliasingLightning.js [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+  if(element) {
+    return element.search('ightning')
+  } else {
+    return false
+  }
+
+}
+
+const replaceWordLightningToThunder = (nodeProperty) => {
```
Comment (seiyria) [inline utils/aliasingLightning.js]: semicolons

File: utils/aliasingLightning.js [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+//   notes: 'Lightning imbue arms and as I type this the only elemental arms you can get. A must keep, even if you don\'t have HMillie these will come in useful later.\n',
+//   type: 'weapon',
+//   subtype: 'arm'
+// }
+// ]
+
+const containsLightning = element => {
+  if(element) {
+    return element.search('ightning')
+  } else {
```
Comment (seiyria) [inline utils/aliasingLightning.js]: this else statement is superfluous; just return false outside of the if

File: utils/aliasingLightning.js [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+    return false
+  }
+
+}
+
+const replaceWordLightningToThunder = (nodeProperty) => {
+  if(nodeProperty) {
+    return nodeProperty.replace(/lightning/gi, 'Thunder')
+  } else {
```
Comment (seiyria) [inline utils/aliasingLightning.js]: this else is also superfluous

File: utils/aliasingLightning.js [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+  }
+
+}
+
+// this function will look up item factors and notes prop
+// if the word "lightning" found then it will replace t to thunderthe returned value is the copy of the item
+// else it will return the item
+
+const aliasingLightning = item => {
+  const {factors} = item
```
Comment (seiyria) [inline utils/aliasingLightning.js]: spacing, semicolons

File: utils/aliasingLightning.js [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+  const [first,second, ...rest] = factors
+  if(
+    containsLightning(item.factors[0].element) ||
+    containsLightning(item.factors[0].desc) ||
+    containsLightning(item.notes)) {
+    return second != undefined ? Object.assign({}, item, {
```
Comment (seiyria) [inline utils/aliasingLightning.js]: this code needs to be cleaned up. don't do != undefined, use strict equality checks when needed. but, in this case, just check if !second.

Also, don't do all of this logic in a return statement. Do it outside of the return.

File: utils/aliasingLightning.js [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+        Object.assign({}, factors.desc, {
+          desc: replaceWordLightningToThunder(first.desc),
+          element: replaceWordLightningToThunder(first.element),
+        }),
+        ...rest
+      ],
+      notes: replaceWordLightningToThunder(item.notes)
+    })
+
+  } else {
```
Comment (seiyria) [inline utils/aliasingLightning.js]: this else statement is superfluous

**Review by seiyria — CHANGES_REQUESTED** [review]
lots of clean up needed here before I could accept it

**Comment by eiyu** [comment]
Ok thanks I learn new things, i've cleaning up and changes the code like your instruction, could you please review my code again?

**Comment by eiyu** [comment]
why the element icon and the lb icon not apear in my copy-bot?
and also what is that redit use for?

**Comment by seiyria** [comment]
The icon doesn't appear because it's looking for an emoji called `element-lightning` or something. So, you'll need to consider that when attempting to display the emoji.

Reddit is so the bot can be pinged on reddit. I believe it's a text-only display.

File: index.js [hunk 16]
```diff
@@ -17,9 +15,12 @@ const { roomInit, room } = require('./commands/room');
 const { contribute } = require('./commands/contribute');
 const { help } = require('./commands/help');
 
+const {aliasLightning} = require('./utils/aliasLightning');
```
Comment (seiyria) [inline index.js]: space after the brace - see the other imports.

also, this should still be in the `shared.js` file.

File: commands/item.js (hunk 3 above)
Comment (seiyria) [reply commands/item.js]: these comments should still be removed

File: commands/item.js (hunk 4 above)
Comment (seiyria) [reply commands/item.js]: remove

File: utils/aliasLightning.js [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+const replaceWordLightningToThunder = (propVal) => {
+  if(propVal) {
+    return propVal.replace(/lightning/gi, 'Thunder');
+  }
+  return propVal;
+}
+
+const aliasLightning = item => {
+  const isLightning = containsLightning(item.factors[0].element) || containsLightning(item.factors[0].desc) || containsLightning(item.notes);
+  const {factors} = item;
```
Comment (seiyria) [inline utils/aliasLightning.js]: brace spacing

File: utils/aliasLightning.js [hunk 18]
```diff
@@ -0,0 +1,53 @@
...
+  const [first,second, ...rest] = factors;
```
Comment (seiyria) [inline utils/aliasLightning.js]: `first, second, ...rest` - correct spacing

File: utils/aliasLightning.js [hunk 19]
```diff
@@ -0,0 +1,53 @@
...
+        ...rest
+      ],
+      notes: replaceWordLightningToThunder(item.notes)
+    });
+  }
+
+  return item;
+}
+
+module.exports = {aliasLightning};
```
Comment (seiyria) [inline utils/aliasLightning.js]: brace spacing

**Review by seiyria — CHANGES_REQUESTED** [review]
still a few things left to do

File: utils/aliasLightning.js (hunk 19 above)
Comment (eiyu) [reply utils/aliasLightning.js]: ok

File: index.js (hunk 16 above)
Comment (eiyu) [reply index.js]: did you mean move the aliaslightning.js to sharedjs?

Comment (seiyria) [reply index.js]: yes. there's no reason to have a utils folder with so few utils at this time.

**Comment by eiyu** [comment]
about the emoji, I'm going to check it first before creating pull req

**Comment by seiyria** [comment]
As long as you make sure the discord bot, when looking up emoji, reverses Thunder to Lightning, it should be fine!

**Comment by eiyu** [comment]
where is that looking up emoji function or code at?

**Comment by seiyria** [comment]
I don't entirely recall. It's probably a helper function in the shared
file. You can search for emoji to find it for sure.

**Comment by eiyu** [comment]
Screen Shot 2019-10-10 at 8 58 10 AM
the weapon icon is also not appear, what am i missing. Is it because my bot is not in soa channel?

**Comment by seiyria** [comment]
Yes, you'd need a copy of all the emoji. It's fine if you don't have it,
just make sure when it tries to get the lightning emoji it does it
correctly.

**Comment by eiyu** [comment]
const getEmoji = (em) => {
  // reverse alias
  if(em == `sbrElThunder`) {
    return emojiHash['Lightning'] || '';
  }
  return emojiHash[em] || '';
};

I make that change to your code is that okay?

**Comment by eiyu** [comment]
const getEmoji = (em) => {
// reverse alias
if(em == sbrElThunder) {
return emojiHash['sbrElLightning'] || '';
}
return emojiHash[em] || '';
};

it should be like that

**Comment by seiyria** [comment]
I would probably do return emojiHash[em.split('Thunder').replace('L
ightning')]

**Comment by eiyu** [comment]
Ok, I'll clean up my code once again. And hopefully it'll work just fine

**Comment by eiyu** [comment]
return emojiHash[em.split('Thunder').replace('L
ightning').join()]

do we need the join method?

**Comment by seiyria** [comment]
Yes.

**Comment by eiyu** [comment]
oh actually no need to split it just call replace lol

**Comment by seiyria** [comment]
Replace takes a regex, split/join is fine enough here. It looks like I put
in that replace on accident though, sorry.

**Comment by eiyu** [comment]
hmm, replace actualy can take string too. So I think it's fine if we write em.replace('Thunder','Lightning')

It's much simple, split seems to much method chaining inside bracket?

**Comment by seiyria** [comment]
yeah replace works too

**Comment by eiyu** [comment]
I have bad writing habbits lol but I'm trying to improve, if there still anything i missed please let me know

**Comment by seiyria** [comment]
sure, lets give it a try.

**Comment by eiyu** [comment]
can I updating the anamnesiac assets? so the community will keep using your updated bot

**Comment by seiyria** [comment]
sure, you absolutely can. please ask in discord and see if someone will help you get started on it, and let me know if I can help any further.

**Pull request merged**